File: lab_7/github_zadania/lab2.cpp
#include <iostream>

using namespace std;

// Open/Closed Principle (OCP) – Obiekty i klasy powinny być otwarte na rozszerzenia, ale zamknięte na modyfikacje.

// Interfejs rabatu
class Discount {
public:
    virtual double calculateDiscount(double price) const = 0; 
    virtual ~Discount() = default;
};

// Regularny rabat (10%)
class RegularDiscount : public Discount {
public:
    double calculateDiscount(double price) const override {
        return price * 0.90;
    }
};

// Sezonowy rabat (20%)
class SeasonalDiscount : public Discount {
public:
    double calculateDiscount(double price) const override {
        return price * 0.80; 
    }
};

void showDiscountedPrice(const Discount& discount, double price) {
    cout << "Original price: " << price << endl;
    cout << "Price after discount: " << discount.calculateDiscount(price) << endl;
}


int main() {
    double price = 10.0;

    RegularDiscount regularDiscount;
    SeasonalDiscount seasonalDiscount;

    cout << "Applying regular discount:" << endl;
    showDiscountedPrice(regularDiscount, price);

    cout << "Applying seasonal discount:" << endl;
    showDiscountedPrice(seasonalDiscount, price);

    return 0;
}

File: lab_7/github_zadania/lab2.java
interface Discount { // poprawnie
    double applyDiscount(double price);
}

// Rabat regularny (10%)
class RegularDiscount implements Discount {
    @Override
    public double applyDiscount(double price) {
        return price * 0.90; // 10% rabatu
    }
}

// Rabat sezonowy (20%)
class SeasonalDiscount implements Discount {
    @Override
    public double applyDiscount(double price) {
        return price * 0.80; // 20% rabatu
    }
}

// Rabat świąteczny (15%)
class HolidayDiscount implements Discount {
    @Override
    public double applyDiscount(double price) {
        return price * 0.85; // 15% rabatu
    }
}

class ShoppingCart {
    private Discount discount; 

    public ShoppingCart(Discount discount) {
        this.discount = discount;
    }

    public double calculatePrice(double price) {
        return discount.applyDiscount(price);
    }
}

public class Main {
    public static void main(String[] args) {
        double originalPrice = 100.0; 

        Discount regularDiscount = new RegularDiscount();
        Discount seasonalDiscount = new SeasonalDiscount();
        Discount holidayDiscount = new HolidayDiscount();

        ShoppingCart regularCart = new ShoppingCart(regularDiscount);
        ShoppingCart seasonalCart = new ShoppingCart(seasonalDiscount);
        ShoppingCart holidayCart = new ShoppingCart(holidayDiscount);

        System.out.println("Cena po rabacie regularnym: " + regularCart.calculatePrice(originalPrice));
        System.out.println("Cena po rabacie sezonowym: " + seasonalCart.calculatePrice(originalPrice));
        System.out.println("Cena po rabacie świątecznym: " + holidayCart.calculatePrice(originalPrice));
    }
}
